File: kakao/2018/kakao2018_4.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int str2time(string time)
{
	return((time[0]-'0')*10+(time[1]-'0'))*60+(time[3]-'0')*10+(time[4]-'0');
}

string solution(int n,int t,int m,vector<string> timetable)
{
	string answer="";
	int ans,num;
	num=timetable.size();
	vector<int> timearr;
	
	for(int i=0;i<num;i++)
		timearr.push_back(str2time(timetable[i]));

	sort(timearr.begin(),timearr.end());

	int drivetime=540,people=0;

	for(int i=0;i<n;i++)
	{
		int seat=m;
		for(int j=people;j<num;j++)
		{
			if(timearr[j]<=drivetime)
			{
				seat--;
				people++;
				if(seat==0)break;
			}
		}
		if(i==n-1)
		{
			if(seat==0)ans=timearr[people-1]-1;
			else ans=drivetime;
		}
		drivetime+=t;
		if(drivetime>=1440)break;
	}
	if((ans/60)<10)answer+="0";
	answer+=to_string(ans/60);
	ans%=60;
	answer+=":";
	if(ans<10)answer+="0";
	answer+=to_string(ans);
	return answer;
}

File: kakao/2018/kakao2018_4.py
def solution(n,t,m,timetable):
    timetable=sorted([int(time[:2])*60+int(time[3:]) for time in timetable])
    for i in range(n):
        l=540+(n-1)*t
        if len(timetable)<m:
            return '%02d:%02d'%(l//60,l%60)
        if i==n-1:
            if timetable[0]>l:return '%02d:%02d'%(l//60,l%60)
            time=timetable[m-1]-1;return '%02d:%02d'%(time//60,time%60)
        for j in range(m-1,-1,-1):
            if timetable[j]<=540+i*t:del timetable[j]